src/types/acls.rs

```diff
@@ -0,0 +1,387 @@
+
+        v.sort_unstable();
+
+        write!(f, "{}", v.into_iter().collect::<String>())
+    }
+}
+
+impl From<HashSet<AclRight>> for AclRights {
+    fn from(hash: HashSet<AclRight>) -> Self {
+        Self { data: hash }
+    }
+}
+
+impl From<Vec<AclRight>> for AclRights {
+    fn from(vec: Vec<AclRight>) -> Self {
+        AclRights {
+            data: vec.into_iter().collect(),
+        }
+    }
+}
+
+impl TryFrom<&str> for AclRights {
+    type Error = AclRightError;
+
+    fn try_from(input: &str) -> Result<Self, Self::Error> {
+        if !input
+            .chars()
+            .all(|c| c.is_ascii_lowercase() || c.is_ascii_digit())
+        {
+            return Err(AclRightError::InvalidRight);
+        }
+
+        Ok(input
+            .chars()
+            .map(|c| c.into())
+            .collect::<HashSet<AclRight>>()
+            .into())
+    }
+}
+
+/// Error from parsing AclRight strings
+#[derive(Debug, Eq, PartialEq)]
+#[non_exhaustive]
+pub enum AclRightError {
+    /// Returned when a non-lower-case alpha numeric is provided in the rights list string.
+    InvalidRight,
+}
+
+impl Display for AclRightError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match *self {
+            AclRightError::InvalidRight => {
+                write!(f, "Rights may only be lowercase alpha numeric characters")
+            }
+        }
+    }
+}
+
+impl std::error::Error for AclRightError {}
+
+/// From [section 3.6 of RFC 4313](https://datatracker.ietf.org/doc/html/rfc4314#section-3.6).
+///
+/// This is a wrapper around a single [`Acl`].
+///
+/// The ACL response from the [`Session::get_acl`] IMAP command
+#[self_referencing]
+pub struct AclResponse {
+    data: Vec<u8>,
+    #[borrows(data)]
+    #[covariant]
+    pub(crate) acl: Acl<'this>,
+}
+
+impl AclResponse {
+    /// Parse the given input into a [`Acl`] response.
+    pub fn parse(
+        owned: Vec<u8>,
+        unsolicited: &mut mpsc::Sender<UnsolicitedResponse>,
+    ) -> Result<Self, Error> {
+        AclResponseTryBuilder {
+            data: owned,
+            acl_builder: |input| {
+                // There should only be ONE single ACL response
+                parse_until_done(input, unsolicited, |response| match response {
+                    Response::Acl(a) => Ok(MapOrNot::Map(Acl {
+                        mailbox: a.mailbox,
+                        acls: a
+                            .acls
+                            .into_iter()
+                            .map(|e| AclEntry {
+                                identifier: e.identifier,
+                                rights: e.rights.into(),
+                            })
+                            .collect(),
+                    })),
+                    resp => Ok(MapOrNot::Not(resp)),
+                })
+            },
+        }
+        .try_build()
+    }
+
+    /// Access to the wrapped [`ListRights`] struct
+    pub fn acl(&self) -> &Acl<'_> {
```

Comment:
Instead of having this method be called something different for each type, how about a method by the same name in each one to access the "inner" type. Maybe it could be called `parsed`, `inner`, or `as_ref`?

Reply:
So, `inner` or `parsed` maybe as it is the object inside the wrap
`as_ref()` somewhat, but it is changing types which is not has as_ref() behaves on other rust structs with that contract.

Though once we get to the quota PR there is one of the Responses there that has two types it returns (a QuotaRoot and an array of Quotas)

In the end, your decision on this.

Reply:
I'm 50/50 on `inner` vs `parsed`. `inner` is more "standard", but `parsed` more accurately reflects what it is. I think either will be fine, so I'll leave it to you :)

For a type that has two "inners", I think it's totally fine for the accessors to be named.

Reply:
I pushed a fixup that uses parsed